File: model/feature_spec.py
"""Single source of truth for the xPts v1 feature contract.

Training (model/) and serving (Plan 3, Deno) MUST build features from these
exact constants so the committed coefficient artifact scores identically in
both places.
"""

MODEL_VERSION = "v1.0.0"

# Exp-decay form window (most-recent prior gameweeks) and decay base.
FORM_WINDOW = 6
DECAY_ALPHA = 0.85

POSITIONS = ["GKP", "DEF", "MID", "FWD"]
QUANTILES = [0.25, 0.50, 0.75]

# Per-GW stats we take an exp-decay form average of. Each yields a
# `form_<stat>` feature. `total_points` last so its form == baseline (c).
FORM_STATS = [
    "expected_goals",
    "expected_assists",
    "expected_goal_involvements",
    "threat",
    "creativity",
    "influence",
    "bps",
    "defensive_contribution",
    "total_points",
]

# Fixed scaling so raw features land in a well-conditioned range for the
# linear model and the dot-product scorer stays trivial to reproduce.
VALUE_SCALE = 10.0      # FPL price is in tenths of a million (e.g. 55 -> 5.5)
STRENGTH_SCALE = 1000.0  # bootstrap team strengths are ~1000-1400

FEATURE_COLUMNS = (
    [f"form_{s}" for s in FORM_STATS]
    + ["xmin", "opp_strength_def", "opp_strength_att", "was_home", "value_scaled"]
)

File: supabase/functions/fpl-project/feature-spec.ts
// TS mirror of model/feature_spec.py — the train/serve feature contract.
// Keep IN SYNC with the Python file; the golden-fixture parity test guards it.
export const MODEL_VERSION = 'v1.0.0';
export const FORM_WINDOW = 6;
export const DECAY_ALPHA = 0.85;
export const POSITIONS = ['GKP', 'DEF', 'MID', 'FWD'] as const;
export const QUANTILES = [0.25, 0.5, 0.75] as const;
export const VALUE_SCALE = 10;
export const STRENGTH_SCALE = 1000;

export const FORM_STATS = [
  'expected_goals',
  'expected_assists',
  'expected_goal_involvements',
  'threat',
  'creativity',
  'influence',
  'bps',
  'defensive_contribution',
  'total_points',
] as const;

export const FEATURE_COLUMNS: string[] = [
  ...FORM_STATS.map((s) => `form_${s}`),
  'xmin',
  'opp_strength_def',
  'opp_strength_att',
  'was_home',
  'value_scaled',
];

// #212 seeding contract. MUST stay byte-identical in meaning to
// model/seed_spec.py — the `seed` block of the parity fixture is the guard.
export const SEED_ROWS = FORM_WINDOW;
export const SEED_DENOMINATOR = 38;
export const SEASON_WEIGHTS = [0.7, 0.3] as const;
export const SEED_DEPTH = SEASON_WEIGHTS.length;
export const NEWCOMER_K = 10;
export const SEED_MODEL_VERSION = 'v1.0.0-seed';
